### ISSUE TITLE
openvino  continued package by pyinstaller  external dll  issue

### ISSUE THREAD

**Issue by hcgprague** [microsoft/onnxruntime#7346]
.\build.bat --config Release --build_wheel --use_openvino  

openvino  continued package by pyinstaller  external dll  issue
------------------

CPU_FP32 --parallel --cmake_generator "Visual Studio 16 2019"

pyinstaller -F  ./webapi.spec 

# -*- mode: python ; coding: utf-8 -*-

import sys
sys.setrecursionlimit(5000)

block_cipher = None

a = Analysis(['webapi.py'],
             pathex=['C:\\Users\\ap01\\Desktop\\X\\colocrbarcode\\win7\\build_win7\\app'],
             binaries=[],
             datas=[
			 ( '.\\CAPI\\_pywrap_tensorflow_internal.pyd', '.\\pywarp'),
           		( '.\\CAPI\\onnxruntime_pybind11_state.pyd', '.\\CAPI\\')
		
		],
             hiddenimports=['sklearn.neighbors.typedefs','numpy.core._dtype_ctypes','cython', 'sklearn' , 'sklearn.utils._cython_blas','keras'],
             hookspath=[],
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher,
             noarchive=False)
pyz = PYZ(a.pure, a.zipped_data,
             cipher=block_cipher)
exe = EXE(pyz,
          a.scripts,
          a.binaries,
          a.zipfiles,
          a.datas,
          [],
          name='webapi',
          debug=False,
          bootloader_ignore_signals=False,
          strip=False,
          upx=False,
          upx_exclude=[],
          runtime_tmpdir=None,
          console=True )

**Describe the bug**
======================================================================
After pyinstaller to package to webapi.exe , it run start with error as 2021-04-14 16:00:19.1906275 [E:onnxruntime:Default, provider_bridge_ort.cc:564 onnxruntime::ProviderSharedLibrary::Ensure] Failed to load library, error code: 126

**Urgency**
normal

**System information**
- OPENVINO : 2021.3
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): win10pro 64bit
- ONNX Runtime installed from (source or binary): py  source package  
- ONNX Runtime version: onnxruntime 1.7.0   , ommxruntime_openvino.whl (build)
- Python version:  python 3.6
- Visual Studio version (if applicable): VS2019    ,  .\build.bat --config Release  --build_wheel --use_openvino CPU_FP32  --parallel  --cmake_generator "Visual Studio 16 2019"
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: NO
- GPU model and memory:
- VC++ 2019

**To Reproduce**
- Describe steps/code to reproduce the behavior.
- Attach the ONNX model to the issue (where applicable) to expedite investigation.

**Expected behavior**

**Screenshots**
 
**Additional context**
Add any other context about the problem here. If the issue is about a particular model, please share the model details as well to facilitate debugging.

**Comment by jywu-msft** [comment]
I built OpenVino EP with latest OnnxRuntime master branch on WIndows10 with OpenVINO 2021.3 and the python bindings worked fine.
Per the OpenVINO instructions, did you run setupvars.bat to set appropriate OpenVINO environment variables
prior to running your python application?

**Comment by hcgprague** [comment]
yes , it works run setupvars.bat to set appropriate OpenVINO environment variables
prior to running your python application  in  .py   .   

But after pyinstaller==4.0  to   be .exe .   it run fail , error message is  :   2021-04-14 16:00:19.1906275 [E:onnxruntime:Default, provider_bridge_ort.cc:564 onnxruntime::ProviderSharedLibrary::Ensure] Failed to load library, error code: 126

**Comment by jywu-msft** [comment]
I think I understand now - you are using pyinstaller to package together all the dependencies for execution.
I suspect the issue is that pyinstaller has no way of determining dll dependencies that are explicitly loaded. 
in this case, onnxruntime_pybind11_state.pyd will explicitly load onnxruntime_providers_shared.dll and onnxruntime_providers_openvino.dll
It is probably not able to copy those dll's over when copying modules/dependencies.

**Comment by goksinan** [comment]
I have a related question. I was able to built onnxruntime with OpenVINO. However, when I tested my C++ project utilizing OpenVINO EP for inference, I got the following error:
 `2021-06-04 08:31:36.6926383 [E:onnxruntime:test, provider_bridge_ort.cc:965 onnxruntime::ProviderLibrary::Get] Failed to load library, error code: 126`
Then, I ran `setupvars.bat` to set the OpenVINO environment variables, and inference worked just fine.
But I want to be able to hand over this project, with all required .dll files, to another person who can run the application on his computer. How do I do that? Without OpenVINO EP, including only *onnxruntime.dll* was enough (copying *onnxruntime_providers_shared.dll* and *onnxruntime_providers_openvino.dll* libraries to target folder did not help).

**Comment by gwang-msft** [comment]
You can use the `--add-binary` of `pyinstller` to include the necessary dlls.
First you need to locate where the ONNX Runtime pypi package is installed, for example `<root of you virtual env>\Lib\site-packages\onnxruntime\capi\onnxruntime_providers_shared.dll`, now you will need to add this to,
- Your pyinstaller command
   `--add-binary="<root of you virtual env>\Lib\site-packages\onnxruntime\capi\onnxruntime_providers_shared.dll;.\onnxruntime\capi"`
- OR add this entry to the spec file
   `binaries=[('<root of you virtual env>\\Lib\\site-packages\\onnxruntime\\capi\\onnxruntime_providers_shared.dll', '.\\onnxruntime\\capi')],`

You can do the similar steps for `onnxruntime_providers_cuda.dll` and other similar dlls.

For more information, see pyinstaller manual, https://pyinstaller.readthedocs.io/en/stable/usage.html#cmdoption-add-binary
